### DISCUSSION TITLE
Deploying Nova 4 via Envoyer

### DISCUSSION THREAD

**Discussion by Synchro** [laravel/nova-issues#4558] [Q&A] [answered]
I'm having difficulty deploying Nova via Laravel Envoyer, on a host managed with Forge. To authenticate the Nova installation there are two mechanisms: one to allow composer to download the package, and a second to enable the package license in Laravel. The first is meant to be managed via an `auth.json` file, which provides composer with the necessary credentials. This file (like anything else containing keys or credentials) should not be stored in git, so needs to be stored somewhere else that is accessible during an Envoyer deployment. If it's added directly to the `current` folder on an Envoyer managed host, it won't work, because the next deploy will create a new folder which doesn't contain it.

The second location for Nova credentials is in a `.env` file, however, although this can contain the same credentials as composer needs, it is not read by composer and so the credentials are not available to it.

Annoyingly, this means that while my app runs perfectly on localhost, and it's easy to deploy in CI, getting it to work on my staging and production systems seems unnecessarily difficult. Should I get Envoyer to read my `.env` file and authorise composer manually? This also seems difficult because Envoyer makes a point of running composer tasks for you, and its hook script is not editable; a hook script added beforehand won't help because it's run in a separate shell and does not share its environment.

How is this supposed to be done?

**Comment by Synchro** [comment]
I've got this working via a manual installation and a deploy hook script. I created the `auth.json` file in the project root (`/home/forge/<site-name>`), then added a script that does this (run as the `forge` user):

    cp {{project}}/auth.json {{release}}/auth.json

and placed that before the stock "Install Composer Dependencies" step.

**Reply by ziming** [reply]
Just put it in ur composer global auth.json if the server simply just host this 1 site

**Comment by crynobone** [comment]
I not 100% familiar with Envoyer but I do believe you can store `auth.json` in the project directory and then use

```
cd {{ project }}
composer install -d {{ release }}
```

Reference: https://getcomposer.org/doc/03-cli.md#global-options

**Comment by davidhemphill** [comment, marked as answer]
In Envoyer, you can create a new hook to authenticate Nova:

![CleanShot 2022-07-20 at 11 55 49](https://user-images.githubusercontent.com/58970/180039839-5a6068d9-b1ad-4c3d-837b-173a7168a67d.png)

Something similar to this:

![CleanShot 2022-07-20 at 11 58 03](https://user-images.githubusercontent.com/58970/180040089-838b19c1-f03b-489e-87ad-232440fc4db2.png)
